File: levels/simulate.js
// TEST FIRST LEVEL

// Import libraries we need.
import { default as contract } from 'truffle-contract';

// Import our contract artifacts and turn them into usable abstractions.
import artefHack_artifacts from '../build/contracts/ArtefHack.json'
import dataProvider_artifacts from '../build/contracts/DataProvider.json'
import roles_artifacts from '../build/contracts/Role.json'

function simulate() {
	var DataProvider = contract(dataProvider_artifacts);
	var ArtefHack = contract(artefHack_artifacts);
	var Roles = contract(roles_artifacts);

	var data = require('./data/test.js')
	var setup = require('./functions/setup.js')
	var simulation = require('./functions/simulation.js')
	var artefHack = require('./functions/artefHack.js')
	var dataprovider = require('./functions/dataProvider.js')

	// setup the test environment
	var accounts, web3;
	web3, accounts = setup.init_variables();
	setup.setup_contracts(ArtefHack, DataProvider, Roles, web3);
	role.setRoles(Roles, data.roles);

	// advertiser variables
	var advertiser = accounts[0];
	var advertiser_available_views = 0;

	// constants
	var NUMBER_OF_DAYS = 100;
	var ADVERTISER_BATCH_BUY_COUNT = 100;
	var PUBLISHER_PRICE = 5;

	// output variables
	var visits = 0;
	var satisfied_visits = 0;
	var publisheds_revenue = 0;
	var ad_views = 0;

	// set up simulation
	var current_users = setup.variable(data.users, 1000);
	var current_contents = setup.variable(data.contents, 100);
	dataprovider.insertUsers(DataProvider, current_users);
	publisher.insertContents(DataProvider, current_contents);

	// test your implementation!
	for(var i=0; i < NUMBER_OF_DAYS; i++){
	  // simulate one day of the platform
	  var daily_visits, daily_satisfied_visits, daily_publisher_revenue, daily_ad_views = simulation.one_day(advertiser, current_users, current_contents);
	  // update output variables
	  visits += daily_visits;
	  satisfied_visits += daily_satisfied_visits;
	  publisheds_revenue += daily_publisher_revenue;
	  ad_views += daily_ad_views;
	}
	setup.log_results(visits, (satisfied_visits/visits)*100, publishers_revenue, ad_views);

}

File: levels/py/simulate.py
import json
from web3 import Web3, HTTPProvider

from functions.setup import *
from functions.simulation import *

provider = HTTPProvider('http://localhost:8545')
web3 = Web3(provider)

accounts = web3.eth.accounts
contract_instances = setup_contract_instances(web3)

# TODO move this to cli argument
FIRST_RUN = True

DAYS_COUNT = 2
STARTING_USERS_COUNT = 50

# setup users and contents
users = load_data("users")
contents = load_data("contents")
users_contents_matrix = load_data("users_contents_matrix")
current_users = setup_users(users, STARTING_USERS_COUNT, 4)
# there are no contents in the platform at the start of the simulation
current_contents = []

# insert users and contents
if FIRST_RUN:
	print('')
	print('\033[1mSetup of the simulation\033[0m')
	print('')

	# we might not make users able to sell their data
	# TODO remove DataProvider.sol?
	# print('Inserting users...')
	# insert_users(
	# 	contract_instances,
	# 	current_users,
	# 	accounts,
	# 	STARTING_USERS_COUNT
	# )
	# print('Done! {} users have been created'.format(STARTING_USERS_COUNT))
	# print('')

	# make contents available in the catalogue
	print('Inserting contents...')
	insert_contents(contract_instances, contents, accounts[0])
	print('Done! {} contents have been created'.format(len(contents)))
	print('')

	print('Setting ArtefHack and Publisher addresses')
	set_stakeholders(contract_instances, accounts)
	print('Done!')
	print('')

	print('Setting roles')
	set_roles(contract_instances, accounts)
	print('Done! Roles have been set for {} addresses'.format(len(accounts)))
	print('')

	print('Setting and funding balances')
	set_balances(contract_instances, accounts)
	print('Done! ArtefHack received 1000 tokens, the Advertiser received 50000 tokens, and each user received 10 tokens')
	print('')


print('')
print('\033[1mStarting the simulation...\033[0m')
print('Target : {} days'.format(DAYS_COUNT))
print('')

visits, satisfied_visits, ad_views = [0] * 3
for i in range(0, DAYS_COUNT):

	print('Day {}'.format(i+1))
	d_visits, s_visits, d_views = one_day(
		contract_instances,
		accounts,
		current_users,
		users,
		current_contents,
		contents,
		users_contents_matrix
	)

	visits += d_visits
	satisfied_visits += s_visits
	ad_views += d_views

log_results(
	visits,
	satisfied_visits,
	get_publisher_revenue(contract_instances, accounts[2]),
	ad_views
)
